help.go

```diff
@@ -155,24 +157,91 @@ func ShowAppHelp(c *Context) (err error) {
 	return nil
 }
 
+var shortFlagRegex = regexp.MustCompile(`^-`)
+
 // DefaultAppComplete prints the list of subcommands as the default app completion method
 func DefaultAppComplete(c *Context) {
-	for _, command := range c.App.Commands {
+	DefaultCompleteWithFlags(nil)(c)
+}
+
+func printCommandSuggestions(commands []Command, writer io.Writer) {
+	for _, command := range commands {
 		if command.Hidden {
 			continue
 		}
 		if os.Getenv("_CLI_ZSH_AUTOCOMPLETE_HACK") == "1" {
 			for _, name := range command.Names() {
-				fmt.Fprintf(c.App.Writer, "%s:%s\n", name, command.Usage)
+				fmt.Fprintf(writer, "%s:%s\n", name, command.Usage)
 			}
 		} else {
 			for _, name := range command.Names() {
-				fmt.Fprintf(c.App.Writer, "%s\n", name)
+				fmt.Fprintf(writer, "%s\n", name)
 			}
 		}
 	}
 }
 
+func cliArgContains(flagName string) bool {
+	for _, name := range strings.Split(flagName, ",") {
+		name = strings.TrimSpace(name)
+		count := utf8.RuneCountInString(name)
+		if count > 2 {
+			count = 2
+		}
+		flag := fmt.Sprintf("%s%s", strings.Repeat("-", count), name)
+		for _, a := range os.Args {
+			if a == flag {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func printFlagSuggestions(lastArg string, flags []Flag, writer io.Writer) {
+	cur := shortFlagRegex.ReplaceAllString(lastArg, "")
+	cur = shortFlagRegex.ReplaceAllString(cur, "")
+	for _, flag := range flags {
+		if bflag, ok := flag.(BoolFlag); ok && bflag.Hidden {
+			continue
+		}
+		for _, name := range strings.Split(flag.GetName(), ",") {
+			name = strings.Trim(name, " ")
+			count := utf8.RuneCountInString(name)
+			if count > 2 {
+				count = 2
+			}
+			if strings.HasPrefix(lastArg, "--") && count == 1 {
+				continue
+			}
+			flagCompletion := fmt.Sprintf("%s%s", strings.Repeat("-", count), name)
+			if strings.HasPrefix(name, cur) && cur != name && !cliArgContains(flag.GetName()) {
+				fmt.Fprintln(writer, flagCompletion)
+			}
+		}
+	}
```

Comment:
This whole flag parseroo is a big ugly and should be responsibility of the flag I guess, but that's a separate battle I guessm

Reply:
I don't think so, parsing flag is generic use function and each flag should not be responsible to implementation. Besides this is not parsing the flag just pattern matching it for auto completion